### PR TITLE
CMR-6739 + CMR-6987

### PR DESCRIPTION
**Pull request by mitchstartzel**
CMR-6987:
* Removed ClientId as a required parameter
* OSDD returned correctly if ClientId = '' (empty string)
* Removed ClientId from OSDD itself in case of missing/empty ClientID
* Added tests for all of the above
* Also updated documentation to reflect the change in requirements

CMR-6739
* Added conditional clientId inclusion to coll-specific OSDDs
* Added tests for this as well

### CODE REVIEW

File: app/views/granules/descriptor_document.xml.erb [hunk 1]
```diff
@@ -15,8 +15,12 @@
 	<os:Contact><%= ENV['contact']%></os:Contact>
 	<os:Url type="application/atom+xml" rel="results"
 	  params:method="GET"
-	  template="<%= ENV['opensearch_url']%>/granules.atom?datasetId={echo:datasetId?}&amp;shortName=<%= @short_name %>&amp;versionId=<%= @version_id %>&amp;dataCenter=<%= @data_center %>&amp;boundingBox={geo:box?}&amp;lat={geo:lat?}&amp;lon={geo:lon?}&amp;radius={geo:radius?}&amp;geometry={geo:geometry?}&amp;placeName={geo:name?}&amp;startTime={time:start?}&amp;endTime={time:end?}&amp;cursor={os:startPage?}&amp;numberOfResults={os:count?}&amp;offset={os:startIndex?}&amp;uid={geo:uid?}&amp;parentIdentifier={eo:parentIdentifier?}&amp;clientId=<%= @client_id_model.clientId %>">
-	  <params:Parameter name="datasetId" uiDisplay="Collection identifier" value="{echo:datasetId}" title="Inventory associated with a dataset expressed as an ID" minimum="0"/>
+		<% if !@client_id_model.clientId.nil? && !@client_id_model.clientId.blank? -%>
```
Comment (littleidiot40) [inline app/views/granules/descriptor_document.xml.erb:18]: Again, I think this should be abstracted behind the model validation.

File: app/views/granules/descriptor_document.xml.erb [hunk 2]
```diff
@@ -15,8 +15,12 @@
...
+			template="<%= ENV['opensearch_url']%>/granules.atom?datasetId={echo:datasetId?}&amp;shortName=<%= @short_name %>&amp;versionId=<%= @version_id %>&amp;dataCenter=<%= @data_center %>&amp;boundingBox={geo:box?}&amp;lat={geo:lat?}&amp;lon={geo:lon?}&amp;radius={geo:radius?}&amp;geometry={geo:geometry?}&amp;placeName={geo:name?}&amp;startTime={time:start?}&amp;endTime={time:end?}&amp;cursor={os:startPage?}&amp;numberOfResults={os:count?}&amp;offset={os:startIndex?}&amp;uid={geo:uid?}&amp;parentIdentifier={eo:parentIdentifier?}&amp;clientId=<%= @client_id_model.clientId %>">
+		<% else %>
+	  	template="<%= ENV['opensearch_url']%>/granules.atom?datasetId={echo:datasetId?}&amp;shortName=<%= @short_name %>&amp;versionId=<%= @version_id %>&amp;dataCenter=<%= @data_center %>&amp;boundingBox={geo:box?}&amp;lat={geo:lat?}&amp;lon={geo:lon?}&amp;radius={geo:radius?}&amp;geometry={geo:geometry?}&amp;placeName={geo:name?}&amp;startTime={time:start?}&amp;endTime={time:end?}&amp;cursor={os:startPage?}&amp;numberOfResults={os:count?}&amp;offset={os:startIndex?}&amp;uid={geo:uid?}&amp;parentIdentifier={eo:parentIdentifier?}">
```
Comment (littleidiot40) [inline app/views/granules/descriptor_document.xml.erb:21]: The bulk of the URL is the same in both if and else. Pull out a variable to handle that.

File: app/views/granules/descriptor_document.xml.erb [hunk 3]
```diff
@@ -47,8 +51,12 @@
 	</os:Url>
 	<os:Url type="text/html" rel="results"
 	  params:method="GET"
-	  template="<%= ENV['opensearch_url']%>/granules.html?datasetId={echo:datasetId?}&amp;shortName=<%= @short_name %>&amp;versionId=<%= @version_id %>&amp;dataCenter=<%= @data_center %>&amp;boundingBox={geo:box?}&amp;lat={geo:lat?}&amp;lon={geo:lon?}&amp;radius={geo:radius?}&amp;geometry={geo:geometry?}&amp;placeName={geo:name?}&amp;startTime={time:start?}&amp;endTime={time:end?}&amp;cursor={os:startPage?}&amp;numberOfResults={os:count?}&amp;offset={os:startIndex?}&amp;uid={geo:uid?}&amp;parentIdentifier={eo:parentIdentifier?}&amp;clientId=<%= @client_id_model.clientId %>">
-	  <params:Parameter name="datasetId" uiDisplay="Collection identifier" value="{echo:datasetId}" title="Inventory associated with a dataset expressed as an ID" minimum="0"/>
+		<% if !@client_id_model.clientId.nil? && !@client_id_model.clientId.blank? -%>
```
Comment (littleidiot40) [inline app/views/granules/descriptor_document.xml.erb:54]: Same comments as above.

File: app/views/granules/mosdac.xml.erb [hunk 4]
```diff
@@ -9,8 +9,11 @@ xmlns:params="http://a9.com/-/spec/opensearch/extensions/parameters/1.0/" 	xmlns
     <os:Tags>ISRO SAC MOSDAC dataset</os:Tags>
     <os:Contact>admin@sac.isro.gov.in</os:Contact>
     <os:Url type="application/atom+xml" rel="results" params:method="GET"
-    template="https://mosdac.gov.in/opensearch/datasets.atom?datasetId=<%= @dataset_id %>&amp;boundingBox={geo:box?}&amp;lat={geo:lat?}&amp;lon={geo:lon?}&amp;radius={geo:radius?}&amp;startTime={time:start?}&amp;endTime={time:end?}&amp;startIndex={os:startPage?}&amp;count={os:count?}&amp;gId={gId?}&amp;clientId={clientId?}">
-
+    <% if !@client_id_model.clientId.nil? && !@client_id_model.clientId.blank? -%>
```
Comment (littleidiot40) [inline app/views/granules/mosdac.xml.erb:12]: Don't we have this problem for all of the CWIC providers?

File: app/views/home/docs.html.erb [hunk 5]
```diff
@@ -77,7 +77,7 @@
     <a class="ext" target="_blank" href='http://www.loc.gov/standards/sru/sru-2-0.html'>OASIS Search Retrieve via URL
       2.0 specification</a>.  When requested, facet information is added to the end of the ATOM feed in order to maintain
       the ATOM response compatibility.
-    The OSDD request requires a clientId query parameter used to uniquely identify the client as illustrated in the
+    While not required, the OSDD request allows for a clientId query parameter used to uniquely identify the client as illustrated in the
```
Comment (littleidiot40) [inline app/views/home/docs.html.erb:80]: Put the word 'recommended' somewhere in here.

**Review by littleidiot40 — CHANGES_REQUESTED** [review]

File: app/controllers/granules_controller.rb [hunk 6]
```diff
@@ -13,7 +13,12 @@ def descriptor_document
     @data_center = params[:dataCenter].blank? ? '{echo:dataCenter?}' : params[:dataCenter]
 
     @type = 'granules'
-    if @client_id_model.valid?
+    if !@client_id_model.clientId.blank? && !@client_id_model.valid?
```
[the remark this answers is not in the document]
Comment (mitchstartzel) [reply app/controllers/granules_controller.rb]: Agreed - this logic makes more sense inside the model.

File: app/views/home/docs.html.erb (hunk 5 above)
Comment (mitchstartzel) [reply app/views/home/docs.html.erb]: Reworked this sentence

File: app/views/granules/descriptor_document.xml.erb (hunk 1 above)
Comment (mitchstartzel) [reply app/views/granules/descriptor_document.xml.erb]: There are two points which need to be reconciled here:

First, by way of changes made in this PR, no clientId is now a valid clientId.
Second, we only want to include clientId in the template if one is supplied.

To reconcile this, the blank? check is necessary. However, the nil? check was abstracted behind the clientId model, so we now check for valid? instead.

File: app/models/collection.rb [hunk 7]
```diff
@@ -131,11 +131,13 @@ def to_open_search_collections(document, hits, params, url, for_atom = true, cwi
           if create_cwic_osdd_link || has_granules
             link_title = "Granule OpenSearch Descriptor Document"
             if Flipper.enabled?(:use_cwic_server)
-              #Rails.logger.info "Flipper says use cwic server!"
```
Comment (littleidiot40) [inline app/models/collection.rb:134]: Remove this commented code

File: app/models/collection.rb [hunk 8]
```diff
@@ -131,11 +131,13 @@ def to_open_search_collections(document, hits, params, url, for_atom = true, cwi
...
-              add_link_as_child(doc, node, "#{Rails.configuration.cwic_granules_osdd_endpoint}opensearch/datasets/#{id}/osdd.xml?clientId=#{params[:clientId]}", 'application/opensearchdescription+xml', NEW_REL_MAPPING[:search], link_title)
+              href_link = "#{Rails.configuration.cwic_granules_osdd_endpoint}opensearch/datasets/#{id}/osdd.xml"
+              !params[:clientId].blank? ? href_link = href_link + "?clientId=#{params[:clientId]}" : nil
+              add_link_as_child(doc, node, href_link, 'application/opensearchdescription+xml', NEW_REL_MAPPING[:search], link_title)
             else
-              #Rails.logger.info "Flipper says DO NOT use cwic server!"
```
Comment (littleidiot40) [inline app/models/collection.rb:137]: Remove this commented code.

**Review by littleidiot40 — CHANGES_REQUESTED** [review]

**Review by littleidiot40 — APPROVED** [review]

File: app/models/collection.rb [hunk 9]
```diff
@@ -131,11 +131,13 @@ def to_open_search_collections(document, hits, params, url, for_atom = true, cwi
           if create_cwic_osdd_link || has_granules
             link_title = "Granule OpenSearch Descriptor Document"
             if Flipper.enabled?(:use_cwic_server)
-              #Rails.logger.info "Flipper says use cwic server!"
-              add_link_as_child(doc, node, "#{Rails.configuration.cwic_granules_osdd_endpoint}opensearch/datasets/#{id}/osdd.xml?clientId=#{params[:clientId]}", 'application/opensearchdescription+xml', NEW_REL_MAPPING[:search], link_title)
+              href_link = "#{Rails.configuration.cwic_granules_osdd_endpoint}opensearch/datasets/#{id}/osdd.xml"
+              !params[:clientId].blank? ? href_link = href_link + "?clientId=#{params[:clientId]}" : nil
```
Comment (ygliuvt) [inline app/models/collection.rb:135]: change this to something like: `href_link = href_link + "?clientId=#{params[:clientId]}" unless params[:clientId].blank?`

File: spec/controllers/granules_controller_spec.rb [hunk 10]
```diff
@@ -33,6 +41,28 @@
...
     end
   end
+  context "with a blank clientId" do
+    it "renders a ccmeo OSDD" do
+      get :descriptor_document, :format => :xml, :params => { :clientId => '', :collectionConceptId => 'C1214603059-SCIOPS' }
+      expect(response.status).to eq(200)
+      expect(response).to render_template("granules/ccmeo.xml.erb")
+    end
+    it "renders an eumetsat OSDD" do
+      get :descriptor_document, :format => :xml, :params => { :clientId => '', :collectionConceptId => 'C1588876552-EUMETSAT' }
```
Comment (ygliuvt) [inline spec/controllers/granules_controller_spec.rb:51]: change one of the clientId to `'        '` for test completeness.

File: app/models/collection.rb [hunk 11]
```diff
@@ -131,11 +131,13 @@ def to_open_search_collections(document, hits, params, url, for_atom = true, cwi
...
+              add_link_as_child(doc, node, href_link, 'application/opensearchdescription+xml', NEW_REL_MAPPING[:search], link_title)
             else
-              #Rails.logger.info "Flipper says DO NOT use cwic server!"
-              add_link_as_child(doc, node, "#{ENV['opensearch_url']}/granules/descriptor_document.xml?collectionConceptId=#{id}&clientId=#{params[:clientId]}", 'application/opensearchdescription+xml', NEW_REL_MAPPING[:search], link_title)
+              href_link = "#{ENV['opensearch_url']}/granules/descriptor_document.xml?collectionConceptId=#{id}"
+              !params[:clientId].blank? ? href_link = href_link + "&clientId=#{params[:clientId]}" : nil
```
Comment (ygliuvt) [inline app/models/collection.rb:139]: same comment as above. the `nil` part is confusing. It is easier to read with `unless` or `if`.

**Review by ygliuvt — APPROVED** [review]
only minor comments. Approved assuming the comments are addressed.

**Pull request merged**